### ISSUE TITLE
PyPy, np.string_ and __radd__

### ISSUE THREAD

**Issue by mattip** [numpy/numpy#9620]
Sorry for the length of this report.

This example will print ``'called __radd__'``
```
from __future__ import print_function

class A(str):
    def __radd__(self, other):
        return 'called __radd__'

a = A('abc')
print('abc' + a)
```
since the ``obj1 + obj2`` will call ``__radd__`` if the objects are subclasses.

``np.string_.__radd__`` is the slot function from ``np.generic``, which in C is ``gentype_add``. But calling ``'abc' + np.string('def')`` will **not** call ``gentype_add`` on CPython. The difference can be seen
by calling ``np.string_('abc').__radd__('def')`` which will call ``gentype_add`` and raise a TypeError looking for the proper UFunc to call.

On PyPy, we **do** call ``gentype_add`` in both cases, which is bad since calling ``gentype_add`` fails.

The problem is that there are two seperate sets of methods on a python object; the c-api c-level function slots and the interpreter object class dictionary. ``__radd__`` lives only in the class dictionary, but the interpreter call to ``binary_op1`` in ``abstract.c`` looks only at the c-level function. There is an implicit assumption, which fails in this case, that the two are equivalent.

I think the best thing to do, since this manifests itself only in this one corner case, is to add a dummy ``__radd__`` function to the ``tp_method`` table This would override the ``__radd__`` in ``np.generic`` and "solve" the problem for PyPy. Trying to special case this in PyPy, or submitting and getting a patch through CPython would be much harder. Any thoughts? If there is no strong opposition I will implement this in a future pull request.

**Comment by FrozenBob** [comment]
The exact reason for the slot/attribute mismatch appears to be because `numpy.string_` multiple-inherits from `str` and `numpy.character` (a use case never officially supported in the Python C API), and it doesn't provide its own `tp_as_sequence` or `tp_as_number`.

`PyType_Ready` ends up copying those *only* from `str`, not `numpy.character`, so `numpy.string_` only gets the `sq_concat` from `str`, not the `nb_add` from `numpy.character`. Lookup for `__radd__` still looks through the whole MRO, though.

Whatever ends up happening here, we shouldn't forget that `numpy.unicode_` also exists and exhibits the same behavior:

```
>>> u'asdf' + numpy.unicode(u'fdsa')
u'asdffdsa'
>>> numpy.unicode_(u'fdsa').__radd__(u'asdf')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: ufunc 'add' did not contain a loop with signature matching types dtyp
e('<U4') dtype('<U4') dtype('<U4')
```

**Comment by mattip** [comment]
fixed in #9628
